presto-main/src/main/java/io/prestosql/operator/CollectOperator.java

```diff
@@ -0,0 +1,151 @@
+ */
+package io.prestosql.operator;
+
+import io.prestosql.spi.Page;
+import io.prestosql.spi.connector.ColumnHandle;
+import io.prestosql.spi.predicate.TupleDomain;
+import io.prestosql.sql.planner.DynamicFilter;
+import io.prestosql.sql.planner.plan.PlanNodeId;
+
+import java.util.concurrent.CompletableFuture;
+
+import static com.google.common.base.Verify.verify;
+import static io.prestosql.SystemSessionProperties.getDynamicFilteringMaxRowCount;
+import static java.util.Objects.requireNonNull;
+
+/// This operator acts as a simple "pass-through" pipe, while saving its input pages.
+/// The collected pages' value are used for creating a run-time filtering constraint (for probe-side table scan in an inner join).
+/// We support only small build-side pages (which should be the case when using "broadcast" join).
+public class CollectOperator
+        implements Operator
+{
+    public static class CollectOperatorFactory
+            implements OperatorFactory
+    {
+        private final int operatorId;
+        private final PlanNodeId planNodeId;
+        private final DynamicFilter dynamicFilter;
+
+        public CollectOperatorFactory(int operatorId, PlanNodeId planNodeId, DynamicFilter dynamicFilter)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.dynamicFilter = dynamicFilter;
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            CompletableFuture<TupleDomain<ColumnHandle>> dynamicPredicateFuture = new CompletableFuture<>();
+            dynamicFilter.addFuture(dynamicPredicateFuture);
+            return new CollectOperator(
+                    driverContext.addOperatorContext(operatorId, planNodeId, CollectOperator.class.getSimpleName()),
+                    dynamicFilter, dynamicPredicateFuture);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+            dynamicFilter.noMoreFutures();
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            throw new UnsupportedOperationException("duplicate() is not supported for CollectOperatorFactory");
+        }
+    }
+
+    private final OperatorContext context;
+    private boolean finished;
+    private Page current;
+    private final DynamicFilter.Collector predicateCollector;
+    private final CompletableFuture<TupleDomain<ColumnHandle>> dynamicPredicateFuture;
+    private long remainingRowCount;
+
+    public CollectOperator(OperatorContext context, DynamicFilter dynamicFilter, CompletableFuture<TupleDomain<ColumnHandle>> dynamicPredicateFuture)
+    {
+        this.context = requireNonNull(context, "context is null");
+        this.predicateCollector = dynamicFilter.collector();
+        this.dynamicPredicateFuture = dynamicPredicateFuture;
+        this.remainingRowCount = getDynamicFilteringMaxRowCount(context.getSession());
```

Comment:
I think this will result in the behavior that if any driver on build side sees more than `dynamic_filtering_max_row_count` rows then we give up. Should we rename the config to `dynamic_filtering_max_per_driver_row_count` to avoid confusion ?

Reply:
Sounds good, changed in 4dd335f.